core/src/SYCL/Kokkos_SYCL_ParallelReduce_MDRange.hpp

```diff
@@ -296,11 +305,12 @@ class Kokkos::Impl::ParallelReduce<CombinedFunctorReducerType,
     // At this point, the reduced value is written to the entry in results_ptr
     // and all that is left is to copy it back to the given result pointer if
     // necessary.
-    if (m_result_ptr && !m_result_ptr_device_accessible) {
-      Kokkos::Impl::DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace,
-                             Kokkos::Experimental::SYCLDeviceUSMSpace>(
-          m_space, m_result_ptr, results_ptr,
-          sizeof(*m_result_ptr) * value_count);
+    if (host_result_ptr) {
+      m_space.fence(
+          "Kokkos::Impl::ParallelReduce<SYCL, MDRangePolicy>::execute: result "
+          "not device-accessible");
+      std::memcpy(m_result_ptr, host_result_ptr,
+                  sizeof(value_type) * value_count);
```

Comment:
Why fence+std::memcpy rather than some SYCL async copy?  Is that because the lock would be released too early?

Reply:
```diff
diff --git a/core/src/SYCL/Kokkos_SYCL_ParallelReduce_Range.hpp b/core/src/SYCL/Kokkos_SYCL_ParallelReduce_Range.hpp
index c6b6fb393..be4fd7d00 100644
--- a/core/src/SYCL/Kokkos_SYCL_ParallelReduce_Range.hpp
+++ b/core/src/SYCL/Kokkos_SYCL_ParallelReduce_Range.hpp
@@ -328,9 +328,10 @@ class Kokkos::Impl::ParallelReduce<CombinedFunctorReducerType,
     // and all that is left is to copy it back to the given result pointer if
     // necessary.
     if (host_result_ptr) {
-      space.fence();
-      std::memcpy(m_result_ptr, host_result_ptr,
-                  sizeof(*m_result_ptr) * value_count);
+      Kokkos::Impl::DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace,
+                             Kokkos::Experimental::SYCLDeviceUSMSpace>(
+          space, m_result_ptr, host_result_ptr,
+          sizeof(*m_result_ptr) * value_count);
     }
 
     return last_reduction_event;
```
is up to 2x slower.

Reply:
Add a comment to that effect

Reply:
e66f21abd9cbe5aac2a676b6b78776cd52cccf6b